frontend/expr_inferrer.py

```diff
@@ -0,0 +1,455 @@
+    """
+    tuple_types = []
+    for elem in node.elts:
+        elem_type = infer(elem, context)
+        tuple_types.append(elem_type)
+
+    return TTuple(tuple_types)
+
+def infer_name_constant(node):
+    """Infer the type of name constants like: True, False, None"""
+    if node.value == True or node.value == False:
+        return TBool()
+    elif node.value is None:
+        return TNone()
+    raise NotImplementedError("The inference for {} is not supported.".format(node.value))
+
+def infer_set(node, context):
+    """Infer the type of a homogeneous set
+
+    Returns: TSet(Type t), where t is the type of the set elements
+    """
+    return TSet(_get_elements_types(node.elts, context))
+
+def _get_stronger_numeric(num1, num2):
+    return num1 if num1.strength > num2.strength else num2
+
+def _infer_add(left_type, right_type):
+    if isinstance(left_type, TNumber) and isinstance(right_type, TNumber):
+        # arithmatic addition
+        return _get_stronger_numeric(left_type, right_type)
+
+    if isinstance(left_type, TSequence) and isinstance(right_type, TSequence):
+        # sequence concatenation
+        if isinstance(left_type, TTuple) and isinstance(right_type, TTuple):
+            new_tuple_types = left_type.types + right_type.types
+            return TTuple(new_tuple_types)
+        if isinstance(left_type, TString) and isinstance(right_type, TString):
+            return TString()
+        if isinstance(left_type, TBytesString) and isinstance(right_type, TBytesString):
+            return TBytesString()
+        if isinstance(left_type, TList) and isinstance(right_type, TList):
+            list_types = UnionTypes()
+            list_types.union(left_type.type)
+            list_types.union(right_type.type)
+            return TList(list_types if len(list_types.types) > 1 else list(list_types.types)[0])
+
+    raise TypeError("Cannot perform operation + on two types {} and {}".format(left_type, right_type))
+
+def _infer_mult(left_type, right_type):
+    if isinstance(left_type, TNumber) and isinstance(right_type, TNumber):
+        return _get_stronger_numeric(left_type, right_type)
+    # TODO handle tuple multiplication
+    if isinstance(left_type, TSequence) and right_type.is_subtype(TInt()):
+        return left_type
+    if isinstance(right_type, TSequence) and left_type.is_subtype(TInt()):
+        return right_type
+
+    raise TypeError("Cannot perform operation * on two types {} and {}".format(left_type, right_type))
+
+def _infer_div(left_type, right_type):
+    if isinstance(left_type, TNumber) and isinstance(right_type, TNumber):
+        return TFloat()
+    raise TypeError("Cannot perform operation / on two types {} and {}".format(left_type, right_type))
+
+def _infer_arithmatic(left_type, right_type):
+    if isinstance(left_type, TNumber) and isinstance(right_type, TNumber):
+        return _get_stronger_numeric(left_type, right_type)
```

Comment:
the type TNumber is not always guaranteed to have the 'strength' attribute. at the moment it is because all of its subtypes have it, but what if at some point we add another subtype which does not have the attribute? should we maybe add default values for strength (and name, since we are at it) in TNumber?